File: Javascript/MergeSort.js
function mergeSort (array) {
    if (array.length < 2) return array
    else {
        let first = array.slice(0, Math.floor(array.length / 2))
        let last = array.slice(Math.floor(array.length / 2))
        first = mergeSort(first)
        last = mergeSort(last)

        let firstCounter = 0
        let lastCounter = 0
        let newArray = []

        while (firstCounter <= first.length - 1 && lastCounter <= last.length - 1) {
            if (first[firstCounter] < last[lastCounter]) {
                newArray.push(first[firstCounter])
                firstCounter += 1
            }
            else {
                newArray.push(last[lastCounter])
                lastCounter += 1
            }
        }

        if (firstCounter <= first.length - 1) newArray.push(...first.slice(firstCounter))
        else if (lastCounter <= last.length - 1) newArray.push(...last.slice(lastCounter))

        return newArray
    }
}

File: Python/MergeSort.py
def mergeSort(array):
    if len(array) < 2:
        # if smallest Possible, hence sorted, return
        return array
    else:
        # break into half then break again
        first = array[:len(array) // 2]
        last = array[(len(array) // 2):]
        first = mergeSort(first)
        last = mergeSort(last)

        firstCounter = 0
        lastCounter = 0
        array = []
        # either first element of one or the other will be the largest so
        # compare and keep appending
        while firstCounter <= len(first) - 1 and lastCounter <= len(last) - 1:
            if first[firstCounter] < last[lastCounter]:
                array.append(first[firstCounter])
                firstCounter += 1
            else:
                array.append(last[lastCounter])
                lastCounter += 1

        # if any are still not appended
        if firstCounter <= len(first) - 1:
            array += first[firstCounter:]

        elif lastCounter <= len(last) - 1:
            array += last[lastCounter:]

        return array
